### ISSUE TITLE
Zarr.jl

### ISSUE THREAD

**Issue by visr** [meggart/Zarr.jl#1]
Hi Fabian,

Cool to see you are working on this. Didn't really go through the code yet, but will do later.

I had also made a small start, but basically the only part that was working was the metadata :)
Since I had so little and wasn't sure I'd go through with it, I didn't announce it anywhere. But here it is:
https://gitlab.com/visr/Zarr.jl

If you are up for it, we can join efforts and compare approaches.

I guess we can just call this Zarr.jl eventually, and rename Zarr.jl to ZarrPyCall or something like that?

**Comment by meggart** [comment]
Hi Martijn, great to hear you are interested in this as well. Regarding the naming, I would prefer to leave the name as it is, since there are considerations on making a zarr C library zarr-developers/zarr#317. So I would like to keep the name `Zarr.jl` reserved for a wrapper around the to-be-implemented C lib. 

Regarding our approaches, it looks like we had a similar structure in mind, although my implementation seems to be a bit more advanced. It was heavily use-case driven, I just had a zarr dataset I needed to access, so I would prefer using your metadata-parsing scheme since it seems to be more principled, while mine is distributed over ZArray.jl and ZGroup.jl. 

Regarding functionality, I think the reading part is more or less working reliably (and much faster than the PyCall-based solution), and I am currently working on methods to generate Zarr arrays. 

I would also like to discuss some package design issues, in particular how to handle C and Fortran syle ordering in an efficient way. 

Do you think the package should live in JuliaGeo for discoverability or would you prefer to first develop in your or my namespace and move later either to JuliaGeo or even to zarr-developers if they are interested?

**Comment by visr** [comment]
Since the spec is quite simple, I think it might be doable and preferable to have a pure julia implementation over a C binding. But this is easy to say of course :). Name is not so important right now.

I was also exploring using Dagger Chunks. Would be nice if we can take advantage of this for a higher level interface and have it integrate nicely with Dagger and JuliaDB for instance.

It's clear that you are further along. So probably best to work in this repo. I don't mind much where it lives, moving to JuliaGeo would be ok, but we can also keep it here for now, or indeed zarr-developers. I can try putting up a PR that merges in my Metadata code + tests. Since you were already using this, should we already care about breaking changes?

The C and Fortran order is indeed a bit tricky since, we cannot (like numpy) flip a switch on our arrays. We do have `PermutedDimsArray`. Also had a discussion about this in ArchGDAL, where we ended up just sticking for GDAL's order (so C style). However this is a bit confusing for interoperability with the rest of the Julia ecosystem.

**Comment by meggart** [comment]
> I was also exploring using Dagger Chunks. Would be nice if we can take advantage of this for a higher level interface and have it integrate nicely with Dagger and JuliaDB for instance.

This would be great, although I don't really understand Dagger right now. Maybe I should try to find some learning resources since the docs are a bit sparse. 

> It's clear that you are further along. So probably best to work in this repo. I don't mind much where it lives, moving to JuliaGeo would be ok, but we can also keep it here for now, or indeed zarr-developers. I can try putting up a PR that merges in my Metadata code + tests. Since you were already using this, should we already care about breaking changes?

This would be a good way forward, let me know if you have any questions. Alternatively I could make an iteration on my code first to at least document the functions a bit to increase readability. Let me know if you have problems. I don't have concerns regarding breaking changes. The interface so far is quite simple and more or less uses `zopen` (we can discuss the name) and `getindex` on groups as well as arrays. So my typical workflow too read zarr arrays would be

````julia
using ZarrNative
a = zopen("/path/to/group")
tair = a["air_temperature"]
data = tair[1:10,1:20,:]
````

As long as something similar is still possible after bikeshedding, I am happy.

> The C and Fortran order is indeed a bit tricky since, we cannot (like numpy) flip a switch on our arrays. We do have PermutedDimsArray. Also had a discussion about this in ArchGDAL (link), where we ended up just sticking for GDAL's order (so C style). However this is a bit confusing for interoperability with the rest of the Julia ecosystem.

The solution I chose for NetCDF.jl (which is also used e.g. by the NetCDF R package) was to pretend from the beginning that the order of dimensions is reversed. NetCDF files have C storage order, so for example ncdump would return `time,lat,lon` as dimensions. When you open the file in Julia with NetCDF.jl the dimension order is presented to the user as `lon,lat,time` and the dimension arguments to all getindex and read operations are reversed while the memory model of Julia fits the one in the file. I kind of liked this model and did not see complaining issues so far.

I my initial Zarr.jl I only had to deal with "C"-style datasets, so I simply ignored "F"-style and did the same like in NetCDF.jl, by just reversing the order once the describing JSON is parsed (see the `reverses` in this line https://github.com/meggart/ZarrNative.jl/blob/master/src/ZArray.jl#L72), and pretend to the user a Julia-like memory layout. 

The main problem is now that zarr wants to be explicit about storage order by providing these options and I am not sure how to deal with this. Probably worth discussing this here https://github.com/zarr-developers/zarr/issues/284 we should post progress there as well.
